serialized_data_interface/__init__.py

```diff
@@ -0,0 +1,133 @@
+from hashlib import md5
+from operator import attrgetter
+
+import requests
+import yaml
+from jsonschema import validate
+from ops.charm import CharmBase
+from ops.framework import EventBase, EventSource, Object, ObjectEvents, StoredState
+
+
+__all__ = ["get_interfaces"]
+
+
+def _get_schema(schema):
+    """Ensures schema is retrieved if necessary, then loads it."""
+
+    if isinstance(schema, str):
+        return yaml.safe_load(requests.get(schema).text)
+
+    return schema
+
+
+def get_interfaces(charm):
+    """Reads metadata.yaml to retrieve schema-checked interface objects."""
+
+    with open("metadata.yaml") as f:
+        metadata = yaml.safe_load(f)
+
+    # Ensure we don't have any duplicate keys across both provides and requires
+    dupes = set.intersection(
+        set(metadata.get("provides", {}).keys()),
+        set(metadata.get("requires", {}).keys()),
+    )
+    assert not dupes, f"Found duplicate keys in provides/requires: {dupes}"
+
+    provides = {
+        name: {
+            version: ProvideAppInterface(charm, name, schema)
+            for version, schema in _get_schema(interface["schema"]).items()
+        }
+        for name, interface in metadata.get("provides", {}).items()
+    }
+
+    requires = {
+        name: {
+            version: RequireAppInterface(charm, name, schema)
+            for version, schema in _get_schema(interface["schema"]).items()
+        }
+        for name, interface in metadata.get("requires", {}).items()
+    }
+
+    return {**provides, **requires}
+
+
+class InterfaceDataChanged(EventBase):
+    pass
+
+
+class RequireInterfaceEvents(ObjectEvents):
+    data_changed = EventSource(InterfaceDataChanged)
+
+
+class RequireAppInterface(Object):
```

Comment:
I'm a little worried about the naming of these classes and the automatic mapping to `provides` and `requires` relation points.

Currently this library supposes that the provide charms will always be the ones sending data and the require ones receiving it. 

In the service mesh interface the charms that require the interface are the ones that are actually sending data. I ended up switching up the classes they use in that [interface](https://github.com/canonical/interface-service-mesh/blob/master/interface_service_mesh.py#L29) to make it work but I don't think that's ideal. If I remember correctly @johnsca had a similar issue in the loadbalancer interface where both endpoints send and receive data.

If we are already making these breaking changes we might as well revisit that part and maybe enable both classes to send and receive data.

Reply:
The request / response pattern is probably the second most common pattern for interfaces (behind one-way data publishing), and there's also potential use-cases for both sides publishing data that isn't a paired request+response.  I was also thinking that it would be good to not require everything to be serialized into a single `data` key; maybe the schema should be for the top-level fields rather than what is serialized into a single field.

This raises the question of scoping, though: should this library support all possible use-cases, or be dedicated to the unidirectional publish pattern with a separate library for the other patterns?  How much of the schema handling logic is common across all patterns and how much is specific to the unidirectional publish pattern?

Reply:
Is there a scenario where `provides` / `requires` are used in a relation and information is sent from both sides? Otherwise, I think it makes sense to think of `provides` as providing relation data, and `requires` as requiring relation data. This would be instead of thinking of it as "provides a service-mesh interface", which as you point out, is kind of weird to fit into the existing provides/requires semantics.

Reply:
> This raises the question of scoping, though: should this library support all possible use-cases, or be dedicated to the unidirectional publish pattern with a separate library for the other patterns?

@johnsca I do think that there's value in having an interface that captures what is (pulling number out of the air) 80% of the use cases. In the Kubeflow charms at least, I don't think there's any relation handling that isn't just "here's the data, do with it as you will".

Reply:
> Is there a scenario where `provides` / `requires` are used in a relation and information is sent from both sides?

Yes, the request response pattern is exactly that sort of scenario.  One side sends a request (e.g., "give me a load balancer with these characteristics"), and the other side sends a response (e.g., "your LB was successfully created and here is the address for it").

And if / when we handle version negotiation, that necessarily means data sent by both sides, as well, although that would probably be handled in a more generic fashion.

> In the Kubeflow charms at least, I don't think there's any relation handling that isn't just "here's the data, do with it as you will".

Yeah, I do think that is the most common pattern, and would cover almost all K8s charm cases, but something like request / response is somewhat common as well, even if the response just comes down success / failure (assuming there's something the requesting charm can potentially change about the request to resolve the issue).